File: net/Scene.cpp

using namespace Engine;

Shapes.Scene.Scene(int nshapes, IShape shapes[]) :
	nshapes(nshapes)
{
	this.shapes = new IShape[nshapes];

	for(int i = 0; i < nshapes; i++)
	{
		this.shapes[i] = shapes[i];
	}
}

Shapes.Scene.~Scene()
{
	delete[] shapes;
}

HitPoint Shapes.Scene.Intersection(Vector start, Vector direction)
{
	HitPoint bestHitPoint = NULL;
	HitPoint hitPoint;

	for(int i = 0; i < nshapes; i++)
	{
		hitPoint = shapes[i].Intersection(start, direction);
		if(hitPoint != NULL && (bestHitPoint ? hitPoint.t < bestHitPoint.t : TRUE))
		{
			delete bestHitPoint;
			bestHitPoint = hitPoint;
		}
		else
		{
			delete hitPoint;
		}
	}

	return bestHitPoint;
}

File: net/Scene.cs
using Engine;

namespace Shapes
{
	public class Scene : IShape
	{
		public readonly IShape[] shapes;

		public Scene(IShape[] shapes)
		{
			this.shapes = shapes;
		}

		public HitPoint Intersection(Vector start, Vector direction)
		{
			HitPoint bestHitPoint = null;
			HitPoint hitPoint;

			for(int i = 0; i < shapes.Length; i++)
			{
				hitPoint = shapes[i].Intersection(start, direction);
				if(hitPoint != null && (bestHitPoint == null || hitPoint.t < bestHitPoint.t))
				{
					bestHitPoint = hitPoint;
				}
			}

			return bestHitPoint;
		}

	}
}
